File: basic_spark/chapter10/count_word.scala
import org.apache.spark.rdd.RDD

// val rootPath: String = _
// val file: String = s"${rootPath}/wikiOfSpark.txt"
val file: String = s"wikiOfSpark.txt"

// 读取文件内容
val lineRDD: RDD[String] = spark.sparkContext.textFile(file)
// 以行为单位做分词
val wordRDD: RDD[String] = lineRDD.flatMap(line => line.split(" "))
 
// 创建单词列表list
val list: List[String] = List("Apache", "Spark")
// 使用list列表对RDD进行过滤
val cleanWordRDD: RDD[String] = wordRDD.filter(word => list.contains(word))
// 把RDD元素转换为（Key，Value）的形式
val kvRDD: RDD[(String, Int)] = cleanWordRDD.map(word => (word, 1))
// 按照单词做分组计数
val wordCounts: RDD[(String, Int)] = kvRDD.reduceByKey((x, y) => x + y)
// 获取计算结果
wordCounts.collect
// Array[(String, Int)] = Array((Apache,34), (Spark,63))

File: basic_spark/chapter08/count_word.py
from pyspark import SparkContext

textFile = SparkContext().textFile("./wikiOfSpark.txt")
wordRDD = textFile.flatMap(lambda line: line.split(" "))
cleanWordRDD = wordRDD.filter(lambda word: word != "")
kvRDD = cleanWordRDD.map(lambda word: (word, 1))
wordCountsRDD = kvRDD.reduceByKey(lambda x, y: x + y)

wordCountsRDD.cache()  # 使用cache算子告知Spark对wordCounts加缓存
wordCountsRDD.count()  # 触发wordCounts的计算，并将wordCounts缓存到内存

wordCount = wordCountsRDD.sortBy(lambda x: x[1], False).take(5)
print(wordCount)

targetPath = "output.txt"
wordCountsRDD.saveAsTextFile(targetPath)
